data_steward/cdr_cleaner/cleaning_rules/base_cleaning_rule.py

```diff
@@ -0,0 +1,248 @@
+"""
+A base class for cleaning rules to extend and implement.
+
+DC-718
+
+This is a base class that all cleaning rules should extend.  This base class
+should contain data relevant to each cleaning rule.  It's attributes and
+functions should be used to ensure that new cleaning rules are added with
+an adequate amount of information.  This is useful for things like, reporting
+features.
+"""
+# Python imports
+import logging
+from abc import ABC, abstractmethod
+from typing import List
+
+# Third party imports
+import googleapiclient
+import oauth2client
+
+# Project imports
+import constants.cdr_cleaner.clean_cdr as cdr_consts
+
+LOGGER = logging.getLogger(__name__)
+
+
+class AbstractBaseCleaningRule(ABC):
+    """
+    Contains attributes and functions relevant to all cleaning rules.
+    """
+    string_list = List[str]
+
+    def __init__(self):
+        """
+        Instantiate a cleaning rule with basic attributes.
+        """
+        super().__init__()
+
+    @abstractmethod
+    def get_query_dictionary_list(self, *args, **keyword_args):
+        """
+        Interface to return a list of query dictionaries.
+
+        :returns:  a list of query dictionaries
+        """
+        pass
+
+    @abstractmethod
+    def print_queries(self, *args, **keyword_args):
+        """
+        Helper function to print the SQL a class geenerates.
+        """
+        pass
+
+
+class BaseCleaningRule(AbstractBaseCleaningRule):
+    """
+    Contains attributes and functions relevant to all cleaning rules.
+    """
+    string_list = List[str]
+    cleaning_class_list = List[AbstractBaseCleaningRule]
+
+    def __init__(self,
+                 jira_issue_numbers: string_list = None,
+                 description: str = None,
+                 affected_datasets: string_list = None,
+                 project_id: str = None,
+                 dataset_id: str = None,
+                 sandbox_dataset_id: str = None,
+                 depends_on: cleaning_class_list = None):
+        """
+        Instantiate a cleaning rule with basic attributes.
+
+        Inheriting classes must set issue numbers, description and affected
+        datasets.  As other tickets may affect the SQL of a cleaning rule,
+        append them to the list of Jira Issues.
+        DO NOT REMOVE ORIGINAL JIRA ISSUE NUMBERS!
```

Comment:
I think this should [also?] be in the parameter documentation (below).

Reply:
sure.  added and updated the existing description.